cmd/expiration-mailer/main.go

```diff
@@ -0,0 +1,216 @@
+	var err error
+	expiresStats := make(map[int]int64, len(m.WarningDays))
+	now := time.Now()
+	// E.g. m.WarningDays = [1, 3, 7, 14] days from expiration
+	for i, expiresIn := range m.WarningDays {
+		left := now
+		if i > 0 {
+			left = left.Add(time.Hour * 24 * time.Duration(m.WarningDays[i-1]))
+		}
+		right := now.Add(time.Hour * 24 * time.Duration(expiresIn))
+
+		m.log.Info(fmt.Sprintf("expiration-mailer: Searching for certificates that expire between %s and %s", left, right))
+		var certs []core.Certificate
+		_, err := m.dbMap.Select(
+			&certs,
+			`SELECT cert.* FROM certificates AS cert JOIN certificateStatus AS cs on cs.serial = cert.serial
+       WHERE cert.expires > :cutoffA AND cert.expires < :cutoffB AND cs.expirationNagsSent < :nags AND cert.status != "revoked"
+       ORDER BY cert.expires ASC`,
+			map[string]interface{}{
+				"cutoffA": left,
+				"cutoffB": right,
+				"nags":    len(m.WarningDays) - i,
+			},
+		)
+		if err != nil {
+			m.log.Err(fmt.Sprintf("expiration-mailer: Error loading certificates: %s", err))
+			continue
+		}
+		if len(certs) > 0 {
+			m.log.Info(fmt.Sprintf("expiration-mailer: Found %d certificates, starting sending messages", len(certs)))
+			for _, cert := range certs {
+				regObj, err := m.dbMap.Get(&core.Registration{}, cert.RegistrationID)
+				if err != nil {
+					return err
+				}
+				reg := regObj.(*core.Registration)
+				parsedCert, err := x509.ParseCertificate(cert.DER)
+				if err != nil {
+					return err
+				}
+				err = m.sendWarning(parsedCert, reg.Contact)
+				if err != nil {
+					return err
+				}
+				expiresStats[expiresIn]++
+
+				// Update CertificateStatus object
+				tx, err := m.dbMap.Begin()
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+
+				csObj, err := tx.Get(&core.CertificateStatus{}, cert.Serial)
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+				certStatus := csObj.(*core.CertificateStatus)
+				certStatus.ExpirationNagsSent = len(m.WarningDays) - i
+
+				_, err = tx.Update(certStatus)
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+
+				err = tx.Commit()
+				if err != nil {
+					// BAD
+					tx.Rollback()
+					return err
+				}
+			}
+			m.log.Info("expiration-mailer: Finished sending messages")
+		}
+	}
+	for k, v := range expiresStats {
+		m.stats.Gauge(fmt.Sprintf("CertificatesExpiringIn.%d-days", k), v, 1.0)
+	}
+
+	return err
+}
+
+func (m *mailer) sendWarning(parsedCert *x509.Certificate, contacts []core.AcmeURL) error {
+	expiresIn := int(parsedCert.NotAfter.Sub(time.Now()).Hours() / 24)
+	emails := []string{}
+	for _, contact := range contacts {
+		if contact.Scheme == "mailto" {
+			emails = append(emails, contact.Opaque)
+		}
+	}
+	if len(emails) > 0 {
+		email := emailContent{
+			ExpirationDate:   parsedCert.NotAfter,
+			DaysToExpiration: expiresIn,
+			CommonName:       parsedCert.Subject.CommonName,
```

Comment:
This is unnecessary. The CommonName will always show up in the DNSNames (and eventually there will be no CommonName).